test/source/hashed_test.cpp

```diff
@@ -0,0 +1,178 @@
+#include <doctest/doctest.h>
+#include <xsparse/levels/hashed.hpp>
+#include <xsparse/levels/dense.hpp>
+#include <xsparse/levels/singleton.hpp>
+
+TEST_CASE("Dense-Hashed")
+{
+    constexpr uintptr_t SIZE0 = 3;
+    constexpr uint8_t ZERO = 0;
+    constexpr uintptr_t SIZE1 = 3;
+    constexpr uintptr_t SIZE2 = 3;
+
+    std::unordered_map<uintptr_t, uintptr_t> const umap1{ { 5, 2 }, { 6, 1 }, { 4, 0 } };
+    std::unordered_map<uintptr_t, uintptr_t> const umap2{ { 2, 5 } };
+    std::unordered_map<uintptr_t, uintptr_t> const umap3{ { 3, 3 }, { 1, 4 }, { 0, 6 } };
+
+    std::vector<std::unordered_map<uintptr_t, uintptr_t>> const crd{ umap1, umap2, umap3 };
+
+    xsparse::levels::dense<std::tuple<>, uintptr_t, uintptr_t> d{ SIZE0 };
+
+    xsparse::levels::hashed<std::tuple<decltype(d)>,
+                            uintptr_t,
+                            uintptr_t,
+                            std::vector<std::unordered_map<uintptr_t, uintptr_t>>>
+        h{ SIZE1, SIZE2, crd };
+
+    uintptr_t l1 = 0;
+    for (auto const [i1, p1] : d.iter_helper(std::make_tuple(), ZERO))
+    {
+        CHECK(l1 == i1);
+        CHECK(l1 == p1);
+        uintptr_t l2 = 0;
+        for (auto const [i2, p2] : h.iter_helper(p1))
+        {
+            CHECK(crd[l1].at(i2) == p2);
+            ++l2;
+        }
+        CHECK(l2 == crd[l1].size());
+        ++l1;
+    }
+    CHECK(l1 == SIZE0);
+}
+
+TEST_CASE("Hashed-Hashed")
+{
+    constexpr uintptr_t SIZE0 = 1;
+    constexpr uintptr_t SIZE1 = 1;
+    constexpr uint8_t ZERO = 0;
+    constexpr uintptr_t SIZE2 = 3;
+    constexpr uintptr_t SIZE3 = 3;
+
+    std::unordered_map<uintptr_t, uintptr_t> const umap1{ { 0, 1 }, { 2, 0 }, { 1, 2 } };
+
+    std::unordered_map<uintptr_t, uintptr_t> const umap2{ { 5, 2 }, { 6, 1 }, { 4, 0 } };
+    std::unordered_map<uintptr_t, uintptr_t> const umap3{ { 2, 5 } };
+    std::unordered_map<uintptr_t, uintptr_t> const umap4{ { 3, 3 }, { 1, 4 }, { 0, 6 } };
```

Comment:
The values of `PK` must be non-overlapping always, like they are here, and every value in the range `0..size_crd` must be present. `0` is included, `size_crd` isn't.

Reply:
Same for all the other tests.

Reply:
Okay, so should the `PK` values go from 0 to 3?

Reply:
I've fixed the `size_crd` for all the test cases.